build_runner/lib/src/server/hot_reload_client.dart

```diff
@@ -2,14 +2,85 @@
 // for details. All rights reserved. Use of this source code is governed by a
 // BSD-style license that can be found in the LICENSE file.
 
+import 'dart:async';
+import 'dart:convert';
 import 'dart:html';
+import 'dart:js';
 
+final _assetsDigestPath = r'$assetDigests';
 final _buildUpdatesProtocol = r'$livereload';
 
-main() {
+JsObject _dartLoader = context['\$dartLoader'];
+JsObject _urlToModuleId = _dartLoader['urlToModuleId'];
+
+List<String> get moduleUrls {
+  JsArray callMethod = context['Array']
+      .callMethod('from', [_urlToModuleId.callMethod('keys', [])]);
+  return List.castFrom(callMethod);
+}
+
+String moduleIdByPath(String path) =>
+    _urlToModuleId.callMethod('get', [window.location.origin + '/' + path])
+        as String;
+
+Future<JsObject> reloadModule(String moduleId) {
+  var completer = Completer<JsObject>();
+  _dartLoader.callMethod('forceLoadModule', [
+    moduleId,
+    JsFunction.withThis((_this, JsObject module) => completer.complete(module))
+  ]);
+  return completer.future;
+}
+
+class ReloadHandler {
+  final String Function(String) _moduleIdByPath;
+  final FutureOr<JsObject> Function(String) _reloadModule;
+  final Map<String, String> _digests;
+
+  ReloadHandler(Map<String, String> this._digests,
+      [this._moduleIdByPath = moduleIdByPath,
+      this._reloadModule = reloadModule]);
+
+  void listener(MessageEvent e) async {
+    Map updatedAssetDigests = json.decode(e.data as String);
+    var moduleIdsToReload = <String>[];
+    for (String path in updatedAssetDigests.keys) {
+      if (_digests[path] == updatedAssetDigests[path]) {
+        continue;
+      }
+      var moduleId = _moduleIdByPath(path);
+      if (_digests.containsKey(path) && moduleId != null) {
+        if (moduleId.endsWith('.ddc')) {
+          moduleId = moduleId.substring(0, moduleId.length - 4);
+        }
+        moduleIdsToReload.add(moduleId);
+      }
+      _digests[path] = updatedAssetDigests[path] as String;
+    }
+    if (moduleIdsToReload.isNotEmpty) {
+      await Future.wait(moduleIdsToReload
+          .map((moduleId) => Future.value(_reloadModule(moduleId))));
+      // TODO Search through dependency graph for true parents
+      var mainModule = await _reloadModule('web/main');
+      mainModule['main'].callMethod('main', []);
+    }
+  }
+}
+
+main() async {
+  var modulePaths = moduleUrls
+      .map((key) => key.replaceFirst(window.location.origin + '/', ''))
+      .toList();
+  var modulePathsJson = json.encode(modulePaths);
+
+  var request = await HttpRequest.request('/$_assetsDigestPath',
+      responseType: 'json', sendData: modulePathsJson, method: 'POST');
+  var digests =
```

Comment:
you can use `.cast<String, String>` on the original map instead which is a bit cleaner:

```dart
var digests = (request.response as Map).cast<String, String>();
```